org.osgi.test.assertj.monitoring/src/main/java/org/osgi/test/assertj/monitoring/internal/MonitoringAssertionImpl.java

```diff
@@ -0,0 +1,297 @@
+/*******************************************************************************
+ * Copyright (c) Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *******************************************************************************/
+
+package org.osgi.test.assertj.monitoring.internal;
+
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicMarkableReference;
+import java.util.function.Predicate;
+
+import org.assertj.core.api.ThrowableAssert.ThrowingCallable;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleEvent;
+import org.osgi.framework.FrameworkEvent;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceEvent;
+import org.osgi.test.assertj.NotPartOfPR.Predicates;
+import org.osgi.test.assertj.monitoring.MonitoringAssertion;
+import org.osgi.test.assertj.monitoring.MonitoringAssertionTimeoutStep;
+
+public class MonitoringAssertionImpl implements MonitoringAssertionTimeoutStep, MonitoringAssertion {
+
+	private ThrowingCallable execute;
+
+	private Predicate<?> innerPredicate = (se) -> false;
+
+	public MonitoringAssertionImpl(ThrowingCallable execute) {
+		this.execute = execute;
+
+	}
+
+	/**
+	 * @param execute   - the action that will be invoked.
+	 * @param predicate - Predicate that would be tested against the Events
+	 * @param timeout   - the time in that the matches must happened
+	 */
+	private static MonitoringResultAssertionImpl call(ThrowingCallable execute, Predicate<?> predicate, int timeout) {
+		BundleContext bc = FrameworkUtil.getBundle(MonitoringResultAssertionImpl.class).getBundleContext();
+
+		EventRecording eventRecording = EventRecording.record(bc, execute, predicate, timeout);
+		return new MonitoringResultAssertionImpl(eventRecording, MonitoringResultAssertionImpl.class);
+	}
```

Comment:
It seems to me that the idea behind `MonitoringResultAssertion` is to monitor events in real time (hence the predicates that have timeouts and suggest waiting). However, `EventRecording.record()` will run the `ThrowingCallable` to completion before it returns, so the entire event sequence will have already played out before the assertion instance is even created.

Reply:
This is what it should do.

Run the `ThrowingCallable`  complete and record everything happened in the `ThrowingCallable` but stop directly after the call if the predicate succeeds (or wait until the timeout)  that the predicate succeeds maybe later.

this way it is possible to do a kind of inactivity check where we do some more complex predicates
```
	/**
	 * Until no more service event within given time is fired.
	 *
	 * @param millis      the time in milliseconds
	 * @param objectClass the objectClass
	 * @return MonitoringAssertTimeoutStep - to define the
	 *         observation-timeout.
	 */
	MonitoringAssertionTimeoutStep untilNoMoreServiceEventOfClassWithin(long millis, Class<?> objectClass);
```